File: 6KYU/Reducing by steps/solution.go
package kata

func Gcdi(a, b int) int {
	if a == 0 {
		if b < 0 {
			b = -b
		}
		return b
	}
	return Gcdi(b%a, a)
}

func Som(x, y int) int {
	return x + y
}

func Maxi(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func Mini(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func Lcmu(x, y int) int {
	gcd := Gcdi(x, y)
	if x < 0 {
		x = -x
	}
	if y < 0 {
		y = -y
	}
	a := (x * y) / gcd
	if a < 0 {
		a = -a
	}
	return a
}

type FParam func(int, int) int

func OperArray(f FParam, arr []int, init int) []int {
	l := len(arr)
	res := []int{}
	c := init
	for i := 0; i < l; i++ {
		a := f(c, arr[i])
		res = append(res, a)
		c = a
	}
	return res
}

File: 6KYU/Reducing by steps/solution.clj
(ns operarray.core)

(defn gcdi [a b]
  (loop [a (Math/abs ^long a)
         b (Math/abs ^long b)]
    (if (zero? b)
      a
      (recur b (mod a b)))))

(defn lcmu [a b]
  (/ (* (Math/abs ^long a) (Math/abs ^long b)) (gcdi a b)))

(defn som [x y]
  (+ x y))

(defn maxi [x y]
  (max x y))

(defn mini [x y]
  (min x y))

(defn oper-array [fct arr init]
  (rest (reductions fct init arr)))

(oper-array lcmu [53 83 54 -58 -20 56 57 10] 53)
